contracts/src/StakeTable.sol

```diff
@@ -191,18 +217,66 @@ contract StakeTable is AbstractStakeTable {
         override
         returns (uint64, uint64)
     {
-        bytes32 hash = _hashBlsKey(blsVK);
-        nodes[hash].balance += amount;
-        return (0, 0);
+        bytes32 key = _hashBlsKey(blsVK);
+        Node memory node = nodes[key];
+
+        // The deposit must come from the node's registered account.
+        if (node.account != msg.sender) {
+            revert CallerIsNotTheNodeOwner();
+        }
+
+        // A node cannot deposit more tokens while it waiting to register,
+        if (currentEpoch() <= node.registerEpoch) {
+            revert PrematureDeposit();
+        }
+
+        // A node cannot deposit more tokens if an exit request is in progress.
+        if (node.exitEpoch != 0) {
+            revert ExitRequestInProgress();
+        }
+
+        node.balance += amount;
+        nodes[key] = node;
```

Comment:
maybe only changes those few fields instead of copying the whole struct?

Reply:
When I try something like this, some test fails:

```solidity
[nix-shell:~/repositories/espressosys/espresso-sequencer]$ git diff
diff --git a/contracts/src/StakeTable.sol b/contracts/src/StakeTable.sol
index 54f2cd4..e24cac8 100644
--- a/contracts/src/StakeTable.sol
+++ b/contracts/src/StakeTable.sol
@@ -235,8 +235,7 @@ contract StakeTable is AbstractStakeTable {
             revert ExitRequestInProgress();
         }
 
-        node.balance += amount;
-        nodes[key] = node;
+        nodes[key].balance += amount;
         SafeTransferLib.safeTransferFrom(ERC20(tokenAddress), msg.sender, address(this), amount);
 
         emit Deposit(_hashBlsKey(blsVK), uint256(amount));
```

```
Failing tests:
Encountered 1 failing test in contracts/test/StakeTable.t.sol:StakeTable_Test
[FAIL. Reason: Assertion failed.] test_Deposits_succeeds() (gas: 467140)
```

Reply:
maybe your test is wrong, I double checked in Remix:

```solidity
pragma solidity ^0.8.0;

contract StakeTable {
    struct Node {
        uint256 registerEpoch;
        uint256 exitEpoch;
    }
    mapping(address => Node) public nodes;
    constructor () {
        nodes[msg.sender].registerEpoch = 5;
        nodes[msg.sender].exitEpoch = 6;
    }

    function increase(address addr) public {
        Node memory n = nodes[addr];
        nodes[addr].registerEpoch = n.exitEpoch + 1;
        nodes[addr].exitEpoch = n.exitEpoch + 2;
    }
}
```
storage is updated as expected.

Reply:
This was a bug on my side. Fixed in 8e9c15a.